doc/tutorial/metaprogramming-arith.md

```diff
@@ -0,0 +1,290 @@
+  | `(fromArith% ($x:arith)) => `(fromArith% $x)
+```
+And some examples:
+
+```lean,ignore
+def foo := fromArith% ("x" :* "y") -- mul
+#print foo
+/-
+def foo : Arith :=
+Arith.Mul (Arith.Symbol "x") (Arith.Symbol "y")
+-/
+
+def bar := fromArith% ("x" :+ "y") -- add
+#print bar
+/-
+def bar : Arith :=
+Arith.Add (Arith.Symbol "x") (Arith.Symbol "y")
+-/
+
+def baz := fromArith% ("x" :+ 20) -- symbol + int
+#print baz
+/-
+def baz : Arith :=
+Arith.Add (Arith.Symbol "x") (Arith.Int 20)
+-/
+```
+
+All the basic examples work. Let's check that we get precedence right as well:
+
+
+```
+def quux := fromArith% ("x" :+ "y" :* "z") -- multiplication binds tighter
+#print quux
+/-
+def quux : Arith :=
+Arith.Add (Arith.Symbol "x") (Arith.Mul (Arith.Symbol "y") (Arith.Symbol "z"))
+-/
+
+def quuz := fromArith% (("x" :+ "y") :* "z") -- brackets change binding order.
+#print quuz
+/-
+def quuz : Arith :=
+Arith.Mul (Arith.Add (Arith.Symbol "x") (Arith.Symbol "y")) (Arith.Symbol "z")
+-/
+```
+
+
+Writing variables as strings, such as `"x"`  gets old; Wouldn't it be so much
+prettier if we could write `x :* y`, and the macro translated this into `Arith.Mul (Arith.Symbol "x") (Arith.Mul "y")`?
+We can do this, and this will be our first taste of manipulating macro variables using `x.getId`
+instead of just directly evaluating them via `$x`.
+We also write a macro rule for `fromArith%` that translates an identifier into
+a string, using `$(Lean.quote (toString x.getId))`.  (TODO: explain what
+`Lean.quote` does):
+
+```lean,ignore
+-- Have to use french quotes («arith») to refer to `arith` as
+-- `arith` is a keyword after declare_syntax_cat.
+-- We declare identifiers as living in the `arith` syntax category, and
+-- prove an interpretation for them with `fromArith%`
+syntax ident : «arith»  
+
+macro_rules
+  | `(fromArith% $x:ident) => `(Arith.Symbol $(Lean.quote (toString x.getId)))
+```
+
+
+Let's test and see that we can now write expressions such as `x :* y` directly instead of having to write `"x" :* "y"`:
+
+```lean,ignore
+def x_ident := fromArith% x 
+#print x_ident
+/-
+def x_ident : Arith :=
+Arith.Symbol "x"
+-/
+
+def x_plus_y : Arith := fromArith% (x :+ y)
+#print x_plus_y
+/-
+def x_plus_y : Arith :=
+Arith.Add (Arith.Symbol "x") (Arith.Symbol "y")
+-/
+```
+
+We now show an unfortunate consequence of the above definitions. Suppose we want to build `(x :+ y) :+ z)`.
+Since we already have defined `x_plus_y` as `x := y`, perhaps we should reuse it! Let's try:
+
+```lean,ignore
+def x_plus_y_plus_z := x_plus_y :+ z
+#print x_plus_y_plus_z
+/-
+def x_plus_y_plus_z : Arith :=
+Arith.Add (Arith.Symbol "x_plus_y") (Arith.Symbol "z")
+-/
+```
+
+Whoops, that didn't work! What happened? Lean treats `x_plus_y` _itself_ as an identifier! So we need to add some syntax
+to be able to "escape" the `fromArith%` context. Let's use the syntax `<[ $e:term ]>` to mean: evaluate `$e` as a real term,
+not an identifier. The macro looks like follows:
```

Comment:
It is [possible, though not trivial](https://github.com/gebner/quote4/commit/c700ec841dc0d6da08eedc3dde7b4d85c3320de5), to reuse Lean's own antiquotation syntax for DSLs. I wonder if we could make that simpler, e.g. by hardcoding support for a new `$e:antiquot` antiquotation kind. /cc @gebner

Reply:
This is actually much easier than it looks like!  The biggest obstacles were issue #494 and figuring out how to set up the syntax so that `q($x)` parses.  If you're looking for things to improve:
1. Matching on antiquotations could be easier, I'm not sure if I like the category annotation (because it usually specifies the category and not a node).  But writing `` `($[$x]:antiquot) => ... `` would be much more ergonomic than what we have right now.
1. These two-part syntaxes (`do`/`←` in particular, but less so with `` `()``/`$`) are in general awkward to extend because a lot of code hard-codes the other part of the syntax.  (Indeed the really horrible nontrivial part of quote4 is the part which extends match...)
1. I was under the impression `$x` and `$(x)` parse to the same syntax, but `$((x))` does not.  This is because `getAntiquotTerm` returns the same result for the first two syntaxes.  It would be great if pattern-matching could distinguish `$x` and `$(x)`.
1. In the same vein, `$x.y` and `$x.1` surprisingly parse differently.
1. I'm not 100% sure that I implemented escaped antiquotations correctly.  There's very little high-level documentation on how to abuse antiquotations so I've been mostly guessing.  Presumably I need to add an `unescapeAntiquot` to support `$$$$$$$$$$$$x`.
1. Speaking of lots of dollars, it might be easier to understand and easier to write macros if they're counted the other way around.  That is, `$$$$$x` generates five nested antiquotation nodes, and you can write a macro like `` `(fromArith% $[$x]:antiquot) => return x``.

Some of these issues could be solved with better documentation so I think it would be good to introduce antiquotations already here.  That is, make it so that `fromArith% $x + 1` works.